// wagmi-private/connectors/walletLink/dist/wagmi-private-connectors-walletLink.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var base = require('../../../dist/base-e2d9120e.cjs.prod.js');
var providers = require('@ethersproject/providers');
var walletlink = require('walletlink');
var utils = require('ethers/lib/utils');
require('eventemitter3');

var _client = /*#__PURE__*/new WeakMap();

var _provider = /*#__PURE__*/new WeakMap();

class WalletLinkConnector extends base.Connector {
  constructor(config) {
    var _window$ethereum;

    super(config);

    base._defineProperty(this, "id", 'walletLink');

    base._defineProperty(this, "name", 'Coinbase Wallet');

    base._defineProperty(this, "ready", typeof window !== 'undefined' && !((_window$ethereum = window.ethereum) !== null && _window$ethereum !== void 0 && _window$ethereum.isCoinbaseWallet));

    base._classPrivateFieldInitSpec(this, _client, {
      writable: true,
      value: void 0
    });

    base._classPrivateFieldInitSpec(this, _provider, {
      writable: true,
      value: void 0
    });

    base._defineProperty(this, "onAccountsChanged", accounts => {
      if (accounts.length === 0) this.emit('disconnect');else this.emit('change', {
        account: utils.getAddress(accounts[0])
      });
    });

    base._defineProperty(this, "onChainChanged", chainId => {
      var id = base.normalizeChainId(chainId);
      var unsupported = this.isChainUnsupported(id);
      this.emit('change', {
        chain: {
          id,
          unsupported
        }
      });
    });

    base._defineProperty(this, "onDisconnect", () => {
      this.emit('disconnect');
    });
  }

  connect() {
    var _this = this;

    return base._asyncToGenerator(function* () {
      try {
        var provider = _this.getProvider();

        provider.on('accountsChanged', _this.onAccountsChanged);
        provider.on('chainChanged', _this.onChainChanged);
        provider.on('disconnect', _this.onDisconnect);
        var accounts = yield provider.enable();
        var account = utils.getAddress(accounts[0]);
        var id = yield _this.getChainId();

        var unsupported = _this.isChainUnsupported(id);

        return {
          account,
          chain: {
            id,
            unsupported
          },
          provider: new providers.Web3Provider(provider)
        };
      } catch (error) {
        if (/user closed modal/i.test(error.message)) throw new base.UserRejectedRequestError();
        throw error;
      }
    })();
  }

  disconnect() {
    var _this2 = this;

    return base._asyncToGenerator(function* () {
      if (!base._classPrivateFieldGet(_this2, _provider)) return;

      var provider = _this2.getProvider();

      provider.removeListener('accountsChanged', _this2.onAccountsChanged);
      provider.removeListener('chainChanged', _this2.onChainChanged);
      provider.removeListener('disconnect', _this2.onDisconnect);
      provider.disconnect();
      provider.close();

      if (typeof localStorage !== 'undefined') {
        var n = localStorage.length;

        while (n--) {
          var key = localStorage.key(n);
          if (!key) continue;
          if (!/-walletlink/.test(key)) continue;
          localStorage.removeItem(key);
        }
      }
    })();
  }

  getAccount() {
    var _this3 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this3.getProvider();

      var accounts = yield provider.request({
        method: 'eth_accounts'
      }); // return checksum address

      return utils.getAddress(accounts[0]);
    })();
  }

  getChainId() {
    var _this4 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this4.getProvider();

      var chainId = base.normalizeChainId(provider.chainId);
      return chainId;
    })();
  }

  getProvider() {
    if (!base._classPrivateFieldGet(this, _provider)) {
      base._classPrivateFieldSet(this, _client, new walletlink.WalletLink(this.options));

      base._classPrivateFieldSet(this, _provider, base._classPrivateFieldGet(this, _client).makeWeb3Provider(this.options.jsonRpcUrl));
    }

    return base._classPrivateFieldGet(this, _provider);
  }

  getSigner() {
    var _this5 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this5.getProvider();

      var account = yield _this5.getAccount();
      return new providers.Web3Provider(provider).getSigner(account);
    })();
  }

  isAuthorized() {
    var _this6 = this;

    return base._asyncToGenerator(function* () {
      try {
        var account = yield _this6.getAccount();
        return !!account;
      } catch (_unused) {
        return false;
      }
    })();
  }

  switchChain(chainId) {
    var _this7 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this7.getProvider();

      var id = utils.hexValue(chainId);

      try {
        yield provider.request({
          method: 'wallet_switchEthereumChain',
          params: [{
            chainId: id
          }]
        });
        var chains = [..._this7.chains, ...base.allChains];
        return chains.find(x => x.id === chainId);
      } catch (error) {
        if (/user rejected signature request/i.test(error.message)) throw new base.UserRejectedRequestError();else throw new base.SwitchChainError();
      }
    })();
  }

}

exports.WalletLinkConnector = WalletLinkConnector;
