// wagmi-private/connectors/walletConnect/dist/wagmi-private-connectors-walletConnect.esm.js
import { m as _checkPrivateRedeclaration, C as Connector, _ as _defineProperty, a as _classPrivateFieldInitSpec, n as normalizeChainId, b as _asyncToGenerator, U as UserRejectedRequestError, e as _classPrivateFieldGet, d as _classPrivateFieldSet, f as allChains, S as SwitchChainError } from '../../../dist/base-3b1382b0.esm.js';
import { Web3Provider } from '@ethersproject/providers';
import WalletConnectProvider from '@walletconnect/ethereum-provider';
import { getAddress, hexValue } from 'ethers/lib/utils';
import 'eventemitter3';

function _classPrivateMethodInitSpec(obj, privateSet) {
  _checkPrivateRedeclaration(obj, privateSet);
  privateSet.add(obj);
}

function _classPrivateMethodGet(receiver, privateSet, fn) {
  if (!privateSet.has(receiver)) {
    throw new TypeError("attempted to get private field on non-instance");
  }

  return fn;
}

var switchChainAllowedRegex = /(rainbow)/i;

var _provider = /*#__PURE__*/new WeakMap();

var _switchChain = /*#__PURE__*/new WeakSet();

class WalletConnectConnector extends Connector {
  constructor(config) {
    super(config);

    _classPrivateMethodInitSpec(this, _switchChain);

    _defineProperty(this, "id", 'walletConnect');

    _defineProperty(this, "name", 'WalletConnect');

    _defineProperty(this, "ready", true);

    _classPrivateFieldInitSpec(this, _provider, {
      writable: true,
      value: void 0
    });

    _defineProperty(this, "onAccountsChanged", accounts => {
      if (accounts.length === 0) this.emit('disconnect');else this.emit('change', {
        account: getAddress(accounts[0])
      });
    });

    _defineProperty(this, "onChainChanged", chainId => {
      var id = normalizeChainId(chainId);
      var unsupported = this.isChainUnsupported(id);
      this.emit('change', {
        chain: {
          id,
          unsupported
        }
      });
    });

    _defineProperty(this, "onDisconnect", () => {
      this.emit('disconnect');
    });
  }

  connect() {
    var _this = this;

    return _asyncToGenerator(function* () {
      try {
        var _provider$connector$p, _provider$connector, _provider$connector$p2;

        var provider = _this.getProvider(true);

        provider.on('accountsChanged', _this.onAccountsChanged);
        provider.on('chainChanged', _this.onChainChanged);
        provider.on('disconnect', _this.onDisconnect);
        var accounts = yield provider.enable();
        var account = getAddress(accounts[0]);
        var id = yield _this.getChainId();

        var unsupported = _this.isChainUnsupported(id); // Not all WalletConnect options support programmatic chain switching
        // Only enable for wallet options that do


        var walletName = (_provider$connector$p = (_provider$connector = provider.connector) === null || _provider$connector === void 0 ? void 0 : (_provider$connector$p2 = _provider$connector.peerMeta) === null || _provider$connector$p2 === void 0 ? void 0 : _provider$connector$p2.name) !== null && _provider$connector$p !== void 0 ? _provider$connector$p : '';
        if (switchChainAllowedRegex.test(walletName)) _this.switchChain = _classPrivateMethodGet(_this, _switchChain, _switchChain2);
        return {
          account,
          chain: {
            id,
            unsupported
          },
          provider: new Web3Provider(provider)
        };
      } catch (error) {
        if (/user closed modal/i.test(error.message)) throw new UserRejectedRequestError();
        throw error;
      }
    })();
  }

  disconnect() {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      var provider = _this2.getProvider();

      yield provider.disconnect();
      provider.removeListener('accountsChanged', _this2.onAccountsChanged);
      provider.removeListener('chainChanged', _this2.onChainChanged);
      provider.removeListener('disconnect', _this2.onDisconnect);
      typeof localStorage !== 'undefined' && localStorage.removeItem('walletconnect');
    })();
  }

  getAccount() {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      var provider = _this3.getProvider();

      var accounts = provider.accounts; // return checksum address

      return getAddress(accounts[0]);
    })();
  }

  getChainId() {
    var _this4 = this;

    return _asyncToGenerator(function* () {
      var provider = _this4.getProvider();

      var chainId = normalizeChainId(provider.chainId);
      return chainId;
    })();
  }

  getProvider(create) {
    if (!_classPrivateFieldGet(this, _provider) || create) _classPrivateFieldSet(this, _provider, new WalletConnectProvider(this.options));
    return _classPrivateFieldGet(this, _provider);
  }

  getSigner() {
    var _this5 = this;

    return _asyncToGenerator(function* () {
      var provider = _this5.getProvider();

      var account = yield _this5.getAccount();
      return new Web3Provider(provider).getSigner(account);
    })();
  }

  isAuthorized() {
    var _this6 = this;

    return _asyncToGenerator(function* () {
      try {
        var account = yield _this6.getAccount();
        return !!account;
      } catch (_unused) {
        return false;
      }
    })();
  }

}

function _switchChain2(_x) {
  return _switchChain3.apply(this, arguments);
}

function _switchChain3() {
  _switchChain3 = _asyncToGenerator(function* (chainId) {
    var provider = this.getProvider();
    var id = hexValue(chainId);

    try {
      yield provider.request({
        method: 'wallet_switchEthereumChain',
        params: [{
          chainId: id
        }]
      });
      var chains = [...this.chains, ...allChains];
      return chains.find(x => x.id === chainId);
    } catch (error) {
      var message = typeof error === 'string' ? error : error === null || error === void 0 ? void 0 : error.message;
      if (/user rejected request/i.test(message)) throw new UserRejectedRequestError();else throw new SwitchChainError();
    }
  });
  return _switchChain3.apply(this, arguments);
}

export { WalletConnectConnector };
